### DISCUSSION TITLE
What is signal type?

### DISCUSSION THREAD

**Discussion by dineshbvadhia** [python/typing#1042] [Q&A] [answered]
The code is to exit a program when ^C is pressed. How to type this code? 
```
import signal

signal.signal(signal.SIGINT, signal_handler)

def signal_handler(signal, frame): -> None
    print('program terminated')
    return
```

**Comment by Akuli** [comment, marked as answer]
In typeshed, `stdlib/signal.pyi` defines `signal.signal()` here: https://github.com/python/typeshed/blob/3a22bf2411c7357bffb2fe1dc675486a8da97177/stdlib/signal.pyi#L68

Your function is passed as a `_HANDLER`, and the definition of `_HANDLER` just above is: `Union[Callable[[int, Optional[FrameType]], Any], int, Handlers, None]`

You are passing in the function, so we can ignore everything except the `Callable`, and your function should match `Callable[[int, Optional[FrameType]], Any]`. This means a function that takes an `int` as first argument, and `Optional[FrameType]` as second argument, and can return anything it wants.

```python3
import typing
from types import FrameType

def signal_handler(signal: int, frame: Optional[FrameType]) -> None:
    print('program terminated')
```

**Comment by dineshbvadhia** [comment]
Thank-you. My first time here: So, is process to check typeshed first? Secondly, are questions like this for General Discussion or Q&A?

**Reply by Akuli** [reply]
Checking typeshed works. If you use a smart enough editor, you will likely see some kind of popup when you e.g. bring your mouse over `signal.signal`, and those typically display information from typeshed. Sometimes that's enough and you don't have to dig into typeshed yourself.

I don't know whether General Discussion or Q&A is more appropriate, but I'd say it doesn't matter much. I think Q&A makes more sense, as it's clearly a question that you expect someone to be able to just answer.

**Reply by hauntsaninja** [reply]
Often times you can just add your best guess at a type, and if it's wrong your type checker hopefully has a friendly error

**Reply by dineshbvadhia** [reply]
Started type checking with vscode but it is good to understand why for the tricky ones. Back to my editor to finish manually :)

**Reply by dineshbvadhia** [reply]
Hi. Back again. The code

```
import signal
import typing
from types import FrameType

signal.signal(signal.SIGINT, signal_handler)    # line 95

def signal_handler(signal: int, frame: Optional[FrameType]) -> None:
    print('program terminated')
```
causes the error:
```
Traceback (most recent call last):
  File "C:\....py", line 95, in <module>
    signal.signal(signal.SIGINT, signal_handler)
NameError: name 'signal_handler' is not defined
```
What is the problem?

**Reply by Gobot1234** [reply]
You are attempting to use access the signal_handler name before it's actually defined. You need to move the signal.signal call to after you define the function.
